experimental/InjectiveResolutions/src/InjectiveResolutions.jl

```diff
@@ -0,0 +1,573 @@
+    j = 0
+    while !points_in_Q(n_bass,P_Q)
+        bass_ = []
+        for a_bass in n_bass
+            push!(bass_,a_bass+c)
+        end
+        n_bass = bass_
+        j = j +1
+    end
+    return j*c
+end
+
+#computes the quotient of a module M by an ideal I, i.e. (0 :_M I)
+#INPUT:     SubquoModule M
+#           Ideal I
+#OUTPUT:    SubquoModule (0 :_M I)
+function mod_quotient(M::SubquoModule,I::Ideal)
+    T = typeof(M[1])
+    M_I,_ = quotient_by_ideal(I) #base_ring(I)/I as module
+    m = M_I[1] #generator of M_I
+    H = hom(M_I,M)
+
+    Q_gens = Vector{T}()
+    for g in gens(H[1])
+        h_g = element_to_homomorphism(g)
+        b_g = h_g(m)
+        push!(Q_gens,b_g)
+    end
+    return sub(M,Q_gens)
+end
+
+#same as mod_quotient except that the result is generated by non-zero elements
+function mod_quotient_2(M::SubquoModule,I::Ideal)
+    T = typeof(M[1])
+    M_I,_ = quotient_by_ideal(I) #base_ring(I)/I as module
+    m = M_I[1] #generator of M_I
+    H = hom(M_I,M)
+
+    Q_gens = Vector{T}()
+    for g in gens(H[1])
+        h_g = element_to_homomorphism(g)
+        b_g = h_g(m)
+        if !is_zero(b_g) && !(-b_g in Q_gens)
+            push!(Q_gens,b_g)
+        end
+    end
+    return sub(M,Q_gens)
+end
+
+#compute saturation of a module M by an ideal, i.e. (0 :_M I^{infty})
+#INPUT:     Submodule M
+#           Ideal I
+#OUTPUT:    SubquoModule (0 :_M I^{infty})
+function mod_saturate(M::SubquoModule,I::Ideal)
+    M_sat,_ = mod_quotient_2(M,I)
+    M_prev = M_sat #previous module quotient
+    i = 2
+    while true
+        M_q,_ = mod_quotient_2(M,I^i)
+        if M_prev == M_q
+            break
+        end
+        M_sat = M_sat + M_q
+        M_prev = M_q #update
+        i = i +1
+    end
+    return M_sat
+end
+
+#Compute k[ZF]-basis as in Algorithm 3.3
+#INPUT:     SubquoModule M = (0 :_M P_F) computed with mod_quotient(_,_)
+#           prime ideal P_F = k{Q\F}
+#OUTPUT:    k[ZF]-basis of localization (0 :_M P_F)[ZF]
+function ZF_basis(M::SubquoModule,PF::Ideal)
+    T = typeof(M[1])
+    N = M
+    h_N = identity_map(N)
+    B = Vector{T}()
+    for g in gens(M)
+        S = sub(N,[h_N(g)]) #submodule of N =(0 :_M P_F)/(y0,...,yn) generated by g
+        if annihilator(S[1]) == PF && !is_zero(S[1])
+        #if is_subset(PF,annihilator(S[1]))
+            push!(B,g)
+        end
+        M_B,_ = sub(M,B)
+        N,h_N = quo(M,M_B)
+        if is_zero(N)
+            break
+        end
+    end
+    return filter(!is_zero,B) 
+end
+
+#Compute coefficients as in algorithm 3.5
+#INPUT:     SubquoModule M
+#           SubquoModule Mxy = (0 :_M P_F)
+#           homomorphism (inclusion) Mxy -> M
+#           k[ZF]-basis computed as in ZF_basis(_,_)
+function coefficients_incl(N::SubquoModule,Np::SubquoModule,incl_Np::SubQuoHom,Bp::Vector,pr::Ideal)
+    R_N = base_ring(N)
+    if is_zero(modulus(R_N))
+        R_poly = R_N
+    else
+        R_poly = base_ring(R_N)
+    end
+    T = typeof(zero(R_N))
```

Comment:
This really should be
```suggestion
    T = elem_type(R_N)
```
also elsewhere in similar situations